### ISSUE TITLE
ANSIBLE0015 False Positive with variable in "with_fileglob"

### ISSUE THREAD

**Issue by luto** [willthames/ansible-lint#207]
# Issue Type
- Bug report

# Ansible and Ansible Lint details

```
ansible 2.1.2.0
ansible-lint 3.3.3
```

- ansible installation method: OS package (arch)
- ansible-lint installation method: pip

# Desired Behaviour

Using a variable like `"/etc/{{ something }}"` in `with_fileglob` should not cause a `ANSIBLE0015`-error, since there are no bare variables here. Instead this line should pass.

# Actual Behaviour (Bug report only)

Playbook:
```yml
- name: foo
  tasks:
    - name: copy scripts
      debug: msg=a
      with_fileglob: "/etc/{{ something }}"
```

ansible-lint output:
```
[ANSIBLE0015] Found a bare variable '/etc/{{ something }}' used in a 'with_fileglob' loop. You should use the full variable syntax ('{{/etc/{{ something }}}}')
test.yml:3
Task/Handler: copy scripts
```

**Comment by willthames** [comment]
I'm not sure this is fixable in ansible-lint itself given ansible-lint's limited understanding of the content of variables - it doesn't know if `with_fileglob: mylist` should check `{{ mylist }}` or for files called `mylist`. At least one (valid) test breaks no matter which way I attempt to fix this!

There is a workaround for playbook authors - ansible-lint handles the following fine: 

```
- name: foo
  tasks:
    - name: copy scripts
      debug: msg=a
      with_fileglob: 
      - "/etc/{{ something }}"
```

**Comment by willthames** [comment]
cc'ing @stpierre and @jzafran who did a lot of the improvement work on this rule, in case they have any thoughts on a reasonable fix.

**Comment by stpierre** [comment]
I'm not sure I see the problem. The snippet that was posted doesn't seem like valid Ansible to me, since `with_fileglob` takes a list, not a string. The bug here, if there is one, is that ansible-lint gives a misleading error, but that code isn't going to work either way.

**Comment by willthames** [comment]
Not really - ansible doesn't *insist* that `with_fileglob` takes a list (neither does it insist with_items takes a list for that matter) - if it's given a single item it gets converted to a list

```
- hosts: localhost
  connection: local

  tasks:

  - debug:
      msg: "{{ item }}"
    with_fileglob: a

  - debug:
      msg: "{{ item }}"
    with_fileglob:
    - a
```

```
$ ansible-playbook test.yml  -v
No config file found; using defaults
 [WARNING]: Host file not found: /etc/ansible/hosts

 [WARNING]: provided hosts list is empty, only localhost is available


PLAY [localhost] ***************************************************************

TASK [setup] *******************************************************************
ok: [localhost]

TASK [debug] *******************************************************************
ok: [localhost] => (item=/home/wthames/tmp/ansible-lint-207/a) => {
    "item": "/home/wthames/tmp/ansible-lint-207/a", 
    "msg": "/home/wthames/tmp/ansible-lint-207/a"
}

TASK [debug] *******************************************************************
ok: [localhost] => (item=/home/wthames/tmp/ansible-lint-207/a) => {
    "item": "/home/wthames/tmp/ansible-lint-207/a", 
    "msg": "/home/wthames/tmp/ansible-lint-207/a"
}

PLAY RECAP *********************************************************************
localhost                  : ok=3    changed=0    unreachable=0    failed=0   
```

**Comment by stpierre** [comment]
Oh, so it's supported (but undocumented) magical behavior? Kinda classic Ansible right there....

I guess we could do a hackish check to see if the string contains `/`, which would catch all of the examples we have thus far and demonstrate that the string isn't a valid variable. It's certainly not perfect, but the cases it would false positive on are pretty dang rare (e.g., `{{ some_var }}foo`, where `some_var` is expected to have a trailing slash). We could also look to see if the string *contains* a valid variable (`.search()` instead of `.match()`), although I'm not sure what the failure modes of that might be right off.

**Comment by willthames** [comment]
ansible-lint has zero understanding of variables (because it doesn't know what hosts a playbook or role is supposed to apply to) - so we can't check for valid variables.

We could say that a glob must contain `*`, `?` or `{{` - not even sure a `/` is a necessary check - but you might have some more thoughts.

**Comment by stpierre** [comment]
Pretty much everything we're talking about here is a hackish workaround, so I'm not sure I have religion on which approach we use. Your approach sounds like a good start, though -- ensuring that the glob *either* looks like a valid glob (contains `*`, `?`, or a `[]` set) or contains a variable (which we can do with `.search()` instead of `.match()`) seems like it would cover the vast majority of use cases.

Seems like we could do:

    _glob_re = re.compile(r'\*|\?|\[.*\]')
    ...
    if (isinstance(varstring, basestring) and
            (not self._glob_re.search(varstring) and
             not self._jinja.search(varstring)):
        return message.format(...)

**Comment by willthames** [comment]
Thanks @stpierre, I'm of the same opinion. Going to change milestone to 3.4.0 now.

**Comment by luto** [comment]
Except for the missing Hosts-list, the code works just fine.

**Comment by luto** [comment]
Thank you! @willthames

**Comment by doteast** [comment]
Shouldn't this apply to with_lines as well?

**Comment by willthames** [comment]
@doteast, I don't follow - if there's a specific issue with `with_lines` can you raise a new issue and I can take a look?
